cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,269 @@
+    CUDF_TRY(gdf_order_by(_keys.begin(), nullptr,
+                          _keys.num_columns(), _key_sorted_order.get(),
+                          &context));
+  } else {  // Pandas style
+    auto keys_row_bitmask = row_bitmask(_keys, _stream);
+
+    gdf_column null_row_representative = *(_keys.get_column(0));
+    null_row_representative.valid =
+        reinterpret_cast<gdf_valid_type*>(keys_row_bitmask.data().get());
+
+    cudf::table keys{_keys};
+    std::vector<gdf_column*> modified_keys(keys.begin(), keys.end());
+    modified_keys[0] = &null_row_representative;
+    cudf::table modified_keys_table(modified_keys.data(),
+                                    modified_keys.size());
+
+    gdf_context temp_ctx;
+    temp_ctx.flag_null_sort_behavior = GDF_NULL_AS_LARGEST;
+
+    CUDF_TRY(gdf_order_by(modified_keys_table.begin(), nullptr,
+                          modified_keys_table.num_columns(),
+                          _key_sorted_order.get(), &temp_ctx));
+
+    CUDF_TRY(gdf_count_nonzero_mask(
+        reinterpret_cast<gdf_valid_type*>(keys_row_bitmask.data().get()),
+        _keys.num_rows(),
+        &_num_keys));
+  }
+
+  return *_key_sorted_order;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_offsets() {
+  if (_group_offsets)
+    return *_group_offsets;
+
+  _group_offsets = std::make_unique<index_vector>(num_keys());
+
+  auto device_input_table = device_table::create(_keys, _stream);
+  auto sorted_order = static_cast<gdf_size_type*>(key_sort_order().data);
+  decltype(_group_offsets->begin()) result_end;
+  auto exec = rmm::exec_policy(_stream)->on(_stream);
+
+  if (has_nulls(_keys)) {
+    auto comp = row_equality_comparator<true>(*device_input_table, true);
+    result_end = thrust::unique_copy(exec,
+      thrust::make_counting_iterator<gdf_size_type>(0),
+      thrust::make_counting_iterator<gdf_size_type>(num_keys()),
+      _group_offsets->begin(), transform_row_eq_comparator<true>{comp, sorted_order});
+  } else {
+    auto comp = row_equality_comparator<false>(*device_input_table, true);
+    result_end = thrust::unique_copy(exec, 
+      thrust::make_counting_iterator<gdf_size_type>(0),
+      thrust::make_counting_iterator<gdf_size_type>(num_keys()),
+      _group_offsets->begin(), transform_row_eq_comparator<false>{comp, sorted_order});
+  }
+
+  gdf_size_type num_groups = thrust::distance(_group_offsets->begin(), result_end);
+  _group_offsets->resize(num_groups);
+
+  return *_group_offsets;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_labels() {
+  if (_group_labels)
+    return *_group_labels;
+
+  // Get group labels for future use in segmented sorting
+  _group_labels = std::make_unique<index_vector>(num_keys());
+
+  auto& group_labels = *_group_labels;
+  auto exec = rmm::exec_policy(_stream)->on(_stream);
+  auto group_labels_ptr = group_labels.data().get();
+  auto group_offsets_ptr = group_offsets().data().get();
+  thrust::for_each_n(exec,
```

Comment:
I think this would be more intuitive and likely more efficient as a `scatter`. 
```suggestion
  thrust::scatter(exec, thrust::make_constant_iterator(1,0), thrust::make_constant_iterator(1, group_offsets().size()), group_offsets().begin(), _group_labels.begin()); 
```